Setup Towncrier for changelog management, remove release-drafter (#2203)
jazzband/pip-tools: changelog.d/README.md

Reviewer: Did you decide not to follow my boilerplate? Any thoughts? Concrete fragment examples to show? These could be a few of the fragments you're adding in this PR copied into this doc.
Also, let's mention the possibility of bylines.

Reviewer: I like this idea. I was mostly writing those explanations in isolation with no feedback loop, having been tired of re-explaining how people should author their notes. And so I've been dragging this document around across projects.
I think it's imperfect, and I'd like to improve it. I've been thinking of how to integrate some ideas from https://redhat-documentation.github.io/supplementary-style-guide/#release-notes and lurking into the release notes channel in the Write The Docs Slack, but haven't gotten to it yet.

So if you manage to make it more friendly, I'd love to copy it back into other projects. Making `sphinxcontrib-towncrier` was one of the sides of me attempting to improve authoring change logs.

> ```
> ### Example Change Notes
> 
> <copy from cheroot example>
> ```

I think you should copy some of the actual change notes you're adding in this PR. I've been doing that, and it feels more immersive when yarl fragment examples talk about yarl and aiohttp examples talk about aiohttp.

I'd be nice to improve my bot eventually to be able to help with change note authoring, similar to CPython's `blurb-it`. But that's a separate endeavor.

## Adding Change Notes with PRs

It is important to maintain a changelog to explain to users what changed
between versions.

To avoid merge conflicts, we use
[Towncrier](https://towncrier.readthedocs.io/en/stable/) to maintain our
changelog.

Towncrier uses separate files, "news fragments", for each pull request.
On release, those fragments are compiled into the changelog.

You don't need to install Towncrier to contribute, you just have to follow some
simple rules!

- In your pull request, add a new file into `changelog.d/` with a filename
  formatted as `$NUMBER.$CATEGORY.md`.

  - The number is the PR number or issue number which your PR addresses.

  - The category is `bugfix`, `feature`, `deprecation`, `breaking`, `doc`,
    `packaging`, `contrib`, or `misc`.

  - For example, if your PR fixes bug #404, the newsfile should be named
    `changelog.d/404.bugfix.md`.

> - If multiple issues are addressed, make an identical copy of the newsfile with

Reviewer: I'd prefer a symlink, though.

  another issue number in the name.
  Towncrier will automatically merge identical files into one entry with multiple links.

- Prefer the simple past or constructions with "now".

> You can preview the changelog by running `tox run -e changelog-draft`.

Reviewer: I would probably tell the users to build the Sphinx docs and see it rendered in the surrounding context.

Reviewer: Yes. Though, that tox env was originally intended for certain GHA integrations IIRC. I haven't asked to implement that in the issue since that would extend the scope and complexity. Perhaps in the future..

## Adding Change Notes with PRs

It is important to maintain a changelog to explain to users what changed
between versions.

To avoid merge conflicts, we use
[Towncrier](https://towncrier.readthedocs.io/en/stable/) to maintain our
changelog.

Towncrier uses separate files, "news fragments", for each pull request.
On release, those fragments are compiled into the changelog.

You don't need to install Towncrier to contribute, you just have to follow some
simple rules!

- In your pull request, add a new file into `changelog.d/` with a filename
  formatted as `$NUMBER.$CATEGORY.md`.

  - The number is the PR number or issue number which your PR addresses.

  - The category is `bugfix`, `feature`, `deprecation`, `breaking`, `doc`,
    `packaging`, `contrib`, or `misc`.

  - For example, if your PR fixes bug #404, the change notes should be named
    `changelog.d/404.bugfix.md`.

- If multiple issues are addressed, create a symlink to the change notes with
  another issue number in the name.

>   Towncrier will automatically merge files into one entry with multiple links.

Reviewer: Should we also mention that if a change set addresses multiple aspects, it's possible to add change notes with multiple categories and the same number? Symlinking also works.

Reviewer: Ack, marking this as resolved.

- Prefer the simple past or constructions with "now".

> - Include a byline, `` `-- by {user}`github-username` ``

Author: Ach! Good catch! I was fiddling with it to get the right rendering and slipped in an extra backtick by mistake.

You can preview the changelog by running `tox run -e build-docs` and viewing
the changelog in the docs.

### Categories

The categories for change notes are defined as follows.

- `bugfix`: A fix for something we deemed improper or undesired behavior.

- `feature`: A new behavior, such as a new flag or environment variable.

- `deprecation`: A declaration of future removals and breaking changes in behavior.

> - `breaking`: A change in behavior which changes or violates established user expectations. e.g., Removing a flag or changing output formatting.

Reviewer: Should this be capitalized differently?
```suggestion
- `breaking`: A change in behavior which changes or violates established user expectations.
  E.g., removing a flag or changing output formatting.
```

- `doc`: Notable updates to the documentation structure or build process.

- `packaging`: Changes in how `pip-tools` itself is packaged and tested which may impact downstreams and redistributors.

- `contrib`: Changes to the contributor experience. e.g., Running tests, building the docs, or setting up a development environment.

- `misc`: Changes that don't fit any of the other categories.

Sometimes it's not clear which category to use for a change.
Do your best and a maintainer can discuss this with you during review.

### Examples

Example bugfix, `2082.bugfix.md`:

```md
Fixed the ordering of format controls to preserve underlying `pip` behavior
-- by {user}`sethmlarson`.
```

> Example contributor update, `2202.contrib.md`:

Reviewer 2: I wonder whether we could/should add fully-qualified GitHub hyperlinks to the numbered pull requests mentioned in these two examples?

Someone attempting to follow the guidance closely might find that helpful to confirm how the number part of the filename corresponds to the origin of the changeset.

Reviewer: I think it's a good idea. And it aligns with my earlier suggestion to use real change pip-tools PR notes in examples.

```md
`pip-tools` now uses [`sphinx-issues`](https://github.com/sloria/sphinx-issues)
to link to issues, PRs, commits, and user accounts
-- by {user}`sirosen`.
```

### Rationale

> When somebody makes a change, they must record the bits that would affect end-users only including information that would be useful to them.

Author: When I first read this, I thought it was introducing the lines below; hence my long-form comment including notes about brevity. But that's a trick of how GitHub shows the diff, which is rather unfortunate.

I think we can improve this phrasing -- and I notice now `affect` vs `effect` -- and at first glance this suggestion seems good to me. I'll circle back on this later and quite possibly accept; thanks!

> When somebody makes a change, they must record the bits that would affect end-users only including information that would be useful to them.

Reviewer 2: Yep, I should perhaps have added an explicit comment/note to explain the suggested subdivision of that paragraph.  Another learning that I'll try to remember for future.

I do think that the existing usage of `affect` is grammatically correct in this context (we are interested in changes that cause an influence - an affect - on the userbase).

Then, when the maintainers publish a new release, they'll automatically use these records to compose a change log for the respective version.

We write change notes in the past tense because this suits the users who will be reading these notes.

> Combined with others, the notes will be a part of the "news digest" telling the readers what **changed** in a specific version of the library since the previous version.

Reviewer: Let's use one of the following:
```suggestion
Combined with others, the notes will be a part of the "news digest" telling the readers what **changed** in a specific version of the app since the previous version.
```
```suggestion
Combined with others, the notes will be a part of the "news digest" telling the readers what **changed** in a specific version of the project since the previous version.
```
```suggestion
Combined with others, the notes will be a part of the "news digest" telling the readers what **changed** in a specific version of `pip-tools` since the previous version.
```

Technically, `pip-tools` isn't meant to be used as an importable library with a public Python API. Saying “library” here would be inconsistent with our vision and might also confuse people into assuming things that aren't necessarily true.

This methodology has several benefits, including those covered by the
[Towncrier Philosophy](https://towncrier.readthedocs.io/en/stable/#philosophy):

> - Change notes separate the user-facing description of changes from the implementation details.

Author: This was discussed above, and we are using the terminology "change notes" intentionally, and trying to be consistent in always calling the newsfiles "change notes".

If you spot us using other terms, that would be appreciated as feedback -- it is possible we have missed some usages.

>   Details go into the git history, but users aren't expected to care about them.

Reviewer 2: Also: is this line missing a hyphen prefix?

- `doc`: Notable updates to the documentation structure or build process.

- `packaging`: Changes in how `pip-tools` itself is packaged and tested which may impact downstreams and redistributors.

- `contrib`: Changes to the contributor experience. e.g., Running tests, building the docs, or setting up a development environment.

- `misc`: Changes that don't fit any of the other categories.

Sometimes it's not clear which category to use for a change.
Do your best and a maintainer can discuss this with you during review.

### Examples

Example bugfix, `2082.bugfix.md`:

```md
Fixed the ordering of format controls to preserve underlying `pip` behavior
-- by {user}`sethmlarson`.
```

Example contributor update, `2202.contrib.md`:

```md
`pip-tools` now uses [`sphinx-issues`](https://github.com/sloria/sphinx-issues)
to link to issues, PRs, commits, and user accounts
-- by {user}`sirosen`.
```

### Rationale

When making a change to `pip-tools`, it is important to communicate the differences that end-users will experience in a manner that they can understand.

> Details of the change that are primarily of interest only to `pip-tools` developers may be irrelevant to most users, and if so, then those details can be omitted from the news fragment.

Reviewer: Do we still want to adjust the terminology here?
```suggestion
Details of the change that are primarily of interest only to `pip-tools` developers may be irrelevant to most users, and if so, then those details can be omitted from the change note.
```

Author: Should be covered by the latest round of tweaks, good eye as always!
